src/privileged/pageMonitor/api.js

```diff
@@ -0,0 +1,363 @@
+"use strict";
+
+/* global AddonManager, ExtensionAPI, ExtensionCommon, ExtensionUtils, Services */
+ChromeUtils.defineModuleGetter(this, "AddonManager", "resource://gre/modules/AddonManager.jsm");
+/* eslint-disable no-var */
+var {AppConstants} = ChromeUtils.import("resource://gre/modules/AppConstants.jsm", {});
+var {Services} = ChromeUtils.import("resource://gre/modules/Services.jsm", {});
+var {EventManager, EventEmitter} = ExtensionCommon;
+var {Management} = ChromeUtils.import("resource://gre/modules/Extension.jsm", null);
+var tabTracker = Management.global.tabTracker;
+/* eslint-enable no-var */
+
+ChromeUtils.defineModuleGetter(
+  this,
+  "BrowserWindowTracker",
+  "resource:///modules/BrowserWindowTracker.jsm",
+);
+
+/** Return most recent NON-PRIVATE browser window, so that we can
+ * manipulate chrome elements on it.
+ */
+function getMostRecentBrowserWindow() {
+  return BrowserWindowTracker.getTopWindow({
+    private: false,
+    allowPopups: false,
+  });
+}
+
+class PageMonitorEventEmitter extends EventEmitter {
+  emitReportBreakage(tabId) {
+    this.emit("report-breakage", tabId);
+  }
+  emitPageBeforeUnload(tabId, data) {
+    this.emit("page-before-unload", tabId, data);
+  }
+  emitPageDOMContentLoaded(tabId, data) {
+    this.emit("page-DOMContentLoaded", tabId, data);
+  }
+  emitErrorDetected(error, tabId) {
+    const recentWindow = getMostRecentBrowserWindow();
+    let uri;
+    let hasException = false;
+    try {
+      uri = Services.io.newURI("https://" + recentWindow.gBrowser.selectedBrowser.currentURI.hostPort);
+      hasException = Services.perms.testExactPermission(uri, "trackingprotection") === Services.perms.ALLOW_ACTION;
+    } catch (e) {
+      // Getting the hostPort for about: and file: URIs fails, but TP doesn't work with
+      // these URIs anyway
+      uri = null;
+    }
+    this.emit("page-error-detected", error, tabId, hasException);
+  }
+  emitExceptionSuccessfullyAdded(tabId) {
+    this.emit("exception-added", tabId);
+  }
+  emitHasException(hasException) {
+    this.emit("has-exception", hasException);
+  }
+  emitPlatform(platform) {
+    this.emit("platform-found", platform);
+  }
+}
+
+/* https://firefox-source-docs.mozilla.org/toolkit/components/extensions/webextensions/functions.html */
+this.pageMonitor = class extends ExtensionAPI {
+  constructor(extension) {
+    super(extension);
+    this.framescriptUrl = extension.getURL("privileged/pageMonitor/framescript.js");
+  }
+
+  onShutdown(shutdownReason) {
```

Comment:
Not sure if this is fired in web extension experiments. Created a follow-up issue: https://github.com/mozilla/CookieRestrictionsBreakageStudy/issues/92